### DISCUSSION TITLE
Discussion: Change Object to be the parent for the other mygame typeclasses?

### DISCUSSION THREAD

**Discussion by Griatch** [evennia/evennia#2567] [Evennia Engine Development]
I wanted to ask for inputs on this.

Another recent discussion thread here highlights a confusion newcomers have to the default inheritance of the base in-world typeclasses (in `mygame/typeclasses/`. This is how the inheritance works today for the in-game classes (inheritance goes left-to-right with parents on the left):

| evennia root | evennia defaults | mygame children |
| --- | --- | --- | 
| DefaultObject |  | Object |
| | DefaultCharacter | Character |
| | DefaultRoom | Room | 
| | DefaultExit | Exit | 

Above, the `Object` on the `mygame` level is _not_ the parent of `Room`/`Character`/`Exit`, whereas on the evennia-level, `DefaultObject` _is_ the parent of all. 

The issue here is that when overriding a feature, one sometimes want that to apply to all types of in-game entities. For example, `DefaultObject` implements a single `return_appearance` hook that is used by all its children. But if wanting to override that on the `mygame` side, there is no single parent class to do so in - one would need to add it to each of the types in turn. 

Now, it's easy enough to change the structure on your own to change this using multiple-inheritance: 
```python
# in e.g. the mygame/typeclasses/characters.py file 
from typeclasses.objects import Object 
class Character(Object, DefaultCharacter):
    pass 
```
If this change were added also to rooms and exits and a change in `Object` will reflect on them. This makes for the following structure: 

| evennia root | evennia defaults | mygame root | mygame children |
| --- | --- | --- | --- |
| DefaultObject |  | Object |
| | DefaultCharacter | | Character |
| | DefaultRoom | |  Room | 
| | DefaultExit | | Exit | 

The question is if this is a structure we should auto-generate for newly initiated game dirs. 

Advantages: 
- Replicating evennia inheritance structure in mygame - the `Object` becomes the parent of all, same as `DefaultObject` is. 
- Easier to override and document override of common hooks - "just put your change in `Object`!"
- Less for the newcomer to change to accomplish a common change to the system. 

Disadvantages: 
- Multiple-inheritance is harder to reason around, so explaining the inheritance tree is harder. 
- No longer any separate base to use for 'things' that are neither characters, rooms or exits.
- It's not hard to make this change yourself (if you are familiar with Python and multi-inheritance, that is...)

Yet another alternative would be to introduce a new `BaseObject` root in mygame and have `Object` inherit directly from that while making it a mixin for the other typeclasses: 

| evennia root | evennia defaults | mygame root | mygame children |
| --- | --- | --- | --- |
| DefaultObject |  | BaseObject |
| |                      |  |  Object | 
| | DefaultCharacter | | Character |
| | DefaultRoom | |  Room | 
| | DefaultExit | | Exit | 

I'm open to suggestions and discussions on this change. This should likely not affect current users all that much, but it may requrie some changes to the settings and what points where by default.

**Comment by kovitikus** [comment]
I do like the 2nd option of `BaseObject` better.

***

> Multiple-inheritance is harder to reason around, so explaining the inheritance tree is harder.

While this is true, it's also important for new folk to understand the inheritance. Establishing it as a common structure in a new project helps introduce it to them, before they get to the point where they are referencing `evennia.objects.objects` module.

***

I personally never had a problem understanding the DefaultObject structure, and its children. But I can see how the new project structure may give the wrong impression.

**Comment by Griatch** [comment]
Any others wanting to give input on this?

**Comment by kenneaal** [comment]
I would say that consistency is the better path. If one can say "Every override typeclass is representative of a core Evennia typeclass", that is more understandable, and you avoid misunderstandings. And also makes the BaseObject a clearly defined class of its own.

**Reply by Griatch** [reply]
@kenneaal 
So option 3 (with the baseobject) then, if I understand you correctly.

**Comment by owllex** [comment]
I'd vote for consistency between evennia and user game inheritance trees by default. The `BaseObject` idea is interesting, but might suffer from a similar kind of confusion if that doesn't map back to a `DefaultBaseObject` or something.

I guess one of the design considerations is how much work `Object` should be doing. Right now, it can be both a Thing/Item/Whatever but is also used to provided lots of inherited functionality to `Characters`, `Exits`, and `Rooms`, and it contains logic specifically to support those inherited types (which sorts of breaks its encapsulation a little bit, though it all works well in practice).

I wonder if it might simplify things from an explanation/understanding perspective if evennia created a new branch off of `Object`, let's call it an `Item` for the sake of discussion, that's specifically meant for tangible things that aren't exits, characters, or rooms. In this world, `Object` could simply be a parent class not intended for instantiation, kind of like how folks don't typically instantiate Python `objects` but instead inherit from them. This also makes it clear how to use `Item` in a way that isn't as clear for `Object`, and gives a sturdy base class for all kinds of custom item logic.

**Reply by Griatch** [reply]
@owllex 

> I wonder if it might simplify things from an explanation/understanding perspective if evennia created a new branch off of Object, let's call it an Item for the sake of discussion, that's specifically meant for tangible things that aren't exits, characters, or rooms. In this world, Object could simply be a parent class not intended for instantiation, kind of like how folks don't typically instantiate Python objects but instead inherit from them. This also makes it clear how to use Item in a way that isn't as clear for Object, and gives a sturdy base class for all kinds of custom item logic.

Am I understanding you correctly, in that you suggest that we go option 1 here, except we make a new `Item` subclass?

**Reply by owllex** [reply]
Yes, that's right.

**Reply by Griatch** [reply]
Thinking about it, would it not become basically case 3 with some renaming then? 
Maybe I still don't follow what you mean, unless you want to make `Object` completely unused. 

| evennia root | evennia defaults | mygame root | mygame children |
| --- | --- | --- | --- |
| DefaultObject |  | Object |
| |                      |  |  Item | 
| | DefaultCharacter | | Character |
| | DefaultRoom | |  Room | 
| | DefaultExit | | Exit |

**Reply by kenneaal** [reply]
If this does become the structure, there should definitely be a DefaultItem in evennia defaults as well, even if it is just a pass class.

**Reply by owllex** [reply]
My thought was to separate the two current uses of `Object` (parent logic and representing in-game things). Use it strictly as a parent class for the other typeclasses (and as a base for future typeclasses) and don't recommend creating one to represent tangible Things -- use the proposed Item/Thing typeclass for that. Then I would create a corresponding `DefaultItem`. So I guess it's a hybrid of the two.

Similar benefits as your first option, but with some different structure with the goal of making it easier to understand why Room/Exit/etc logic is in Object even though Objects can be non-Rooms/Exits.

**Reply by Griatch** [reply]
Ok, thanks for clarifying. :)

**Comment by strikaco** [comment]
I like option 2 for its simplicity. Multiple inheritance is overrated anyway.

**Comment by Griatch** [comment]
Thanks for the discussion and suggestions; in the end I found that changing the actual inheritance structure (like adding an `Item` or change things to inherit from `Object` directly) would break too much backwards compatibility. So I went with a new `ObjectParent` mixin being generated by the default gamedir template (so a multi-inheritance solution). It is means no backwards incompatibility and people with existing game dirs can easily do the same thing if they prefer.
